### PR TITLE
Healthchecks and termination locks

### PR DESCRIPTION
**Pull request by bassjacob**
Add concept of hooks to steveo to allow process to check the environment health and terminate if required

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -41,7 +43,7 @@ class Steveo implements ISteveo {
   }
 
   runner() {
-    return runner(this.config.engine, this.config, this.registry, this.pool, this.logger);
+    return runner(this.config.engine, this.config, this.registry, this.pool, this.logger, this.hooks);
```
Comment (danielmoi) [inline src/index.js]: and now we are 6

File: src/base/base_runner.js [hunk 2]
```diff
@@ -4,13 +4,55 @@ import shuffle from 'lodash.shuffle';
...
   logger: Logger;
 
-  getActiveSubsciptions(topics: Array < string > = null): Array < string > {
+  constructor(hooks: Hooks = {}) {
+    this.errorCount = 0;
+    this.preProcess = hooks.preProcess || (() => Promise.resolve());
+    this.healthCheck = hooks.healthCheck || (() => Promise.resolve());
+    this.terminationCheck = hooks.terminationCheck || (() => Promise.resolve(false));
+    /* eslint-disable no-console */
+    this.logger = {
```
Comment (johndagostino) [inline src/base/base_runner.js]: do we need to keep the logging?

Comment (bassjacob) [reply src/base/base_runner.js]: @johndagostino just makes it a little smoother in regards to typing. no hard requirement. at the moment we're using base_runner as an abstract class, but maybe it should just be an interface or something?

File: src/base/base_runner.js [hunk 3]
```diff
@@ -4,13 +4,55 @@ import shuffle from 'lodash.shuffle';
...
+      debug: console.log.bind(console),
+      info: console.log.bind(console),
+      trace: console.log.bind(console),
+      error: console.log.bind(console),
+    };
+    /* eslint-enable no-console */
+  }
+
+  async checks(onFail) {
+    if (await this.terminationCheck()) {
```
Comment (anupvarghese) [inline src/base/base_runner.js]: what will happen if this throws?

Comment (bassjacob) [reply src/base/base_runner.js]: Good question. Let me see if I can get it set up to not be an issue.

File: src/base/base_runner.js (hunk 2 above)
Comment (johndagostino) [reply src/base/base_runner.js]: @bassjacob nullLogger is already in the package?

File: src/base/base_runner.js [hunk 4]
```diff
@@ -1,16 +1,52 @@
...
   registry: IRegistry;
   logger: Logger;
 
-  getActiveSubsciptions(topics: Array < string > = null): Array < string > {
+  constructor(hooks: Hooks = {}) {
+    this.errorCount = 0;
+    this.preProcess = hooks.preProcess || (() => Promise.resolve());
+    this.healthCheck = hooks.healthCheck || (() => Promise.resolve());
+    this.terminationCheck = hooks.terminationCheck || (() => Promise.resolve(false));
+    /* eslint-disable no-console */
```
Comment (anupvarghese) [inline src/base/base_runner.js]: remove this eslint

**Pull request merged**